### PR TITLE
Dynamic mountpoint detection

### PR DESCRIPTION
**Pull request by vojtechtrefny**
Dynamic detection of mountpoints for existing devices.

blivet.mounts.MountCache parses /proc/mounts and /proc/self/mountinfo to get information about mounted devices. See https://github.com/rhinstaller/blivet/issues/13 for more information.

### CODE REVIEW

**Comment by mulkieran** [comment]
A couple things that strike me at first glance are:
* I think mountCache should be in FS not DeviceFormat. AFAICT, the only thing about mounting in DeviceFormat is mountable, and there is a ton of stuff about mounting in FS.
* When tasks infrastructure goes in, MountsCache should be factored out into an always available Task and the rest of it. Please take a look at https://github.com/mulkieran/blivet/tree/master-tasks and work out how you would do that, and send me a link for that branch. A PR would probably just make the list too busy.
* Can you come up with a way to test these changes?

- mulhern

**Comment by vpodzime** [comment]
> I think mountCache should be in FS not DeviceFormat. AFAICT, the only thing about mounting in DeviceFormat is mountable, and there is a ton of stuff about mounting in FS.

If I understand it correctly, having mountCache in FS not in DeviceFormat would require all the code that tries to get/set it on a ``device.format`` check if that format is a ``FS`` first. I think not having to do this checks is a nice benefit of having it one level higher in the class hierarchy.

> When tasks infrastructure goes in, MountsCache should be factored out into an always available Task and the rest of it. Please take a look at https://github.com/mulkieran/blivet/tree/master-tasks and work out how you would do that, and send me a link for that branch. A PR would probably just make the list too busy.

I think the priority of this depends on in which order we want to merge this and the tasks infrastructure. If this is the first thing to go in, we should focus on merging this and then on "rebasing" the tasks infrastructure.

**Comment by vpodzime** [comment]
This looks good to me. I'm not giving ACK right now because I think this needs some more attention of other members of the team.

File: blivet/util.py [hunk 1]
```diff
@@ -357,6 +358,18 @@ def insert_colons(a_string):
     else:
         return suffix
 
+def md5_file(filename):
+
+    f = open(filename, "rb")
```
Comment (mulkieran) [inline blivet/util.py]: We usually use the file as a context manager these days:

```
md5 = hashlib.md5()
with open(filename, "rb") as f:
    # do your stuff
```

File: blivet/mounts.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+    def _resolveFormat(self, devspec, subvolspec=None):
+
+        for fmt in self.mountpoints.keys():
+            if fmt[0] == devspec:
+                if not fmt[1]:
+                    return fmt
+
+                elif fmt[1] == subvolspec:
+                    return fmt
```
Comment (mulkieran) [inline blivet/mounts.py]:
```
return next((f for (d, s) in self.mountpoints.keys() if d == devspec and (not s or s == subvolspec)), None)
```
should do the same thing and also explicitly returns None.

This method is probably worth a doc string.

**Comment by mulkieran** [comment]
There's a method, util.get_mount_paths() which blivet does not use, but anaconda does use. Would it be possible to use your MountsCache to re-implement that method as well? It seems like that method is duplicating the reading and using the information that MountsCache should have.

File: blivet/mounts.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+    def getMountpoint(self, devspec, subvolspec=None):
+        """ Get mountpoint for selected device
+
+            :param devscpec: device specification, eg. "/dev/vda1"
+            :type devspec: str
+            :param subvolspec: btrfs subvolume specification, eg. ID or name
+            :type subvolspec: str
+            :returns: mountpoint (path)
+            :rtype: str
```
Comment (mulkieran) [inline blivet/mounts.py]:
```
:rtype: str or NoneType
```

File: blivet/mounts.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+        """
+
+        self._cacheCheck()
+
+        if (devspec, subvolspec) not in self.mountpoints.keys():
+            return None
+
+        else:
+            return self.mountpoints[(devspec, subvolspec)]
```
Comment (mulkieran) [inline blivet/mounts.py]:
```
return self.mountpoints.get((devspec, subvolspec))
```
can replace the if statement entirely.

**Pull request closed without merging**